### PR TITLE
RFC: movable array iterators

### PR DESCRIPTION
**Pull request by ishitatsuyuki**
[Rendered](https://github.com/ishitatsuyuki/rfcs/blob/array_into_iter_move/text/0000-array_into_iter_move.md)

### CODE REVIEW

**Comment by SimonSapin** [comment]
I very much want this, but what is return type of `[T; N]::into_iter`? I think it should be `IntoIter<T, N>`, with [const generics](https://github.com/rust-lang/rust/issues/44580). So this needs to wait for that to be implemented.

Or maybe `IntoIter<A> where A: Unsize<[T]>` might work, but stabilizing with `Unsize` itself unstable doesn’t feel right.

----

> Transmute the array so it has ManuallyDrop<T> signature. This should be memory safe.

I believe that in a method taking `self`, the array can be safely moved into `ManuallyDrop::new(…)` without transmute. Hopefully this move optimizes to a no-op?

**Comment by ishitatsuyuki** [comment]
Please prefer line comments so I can respond without copying :)

> I believe that in a method taking self, the array can be safely moved into ManuallyDrop::new(…) without transmute. Hopefully this move optimizes to a no-op?

I think there's a need to be able to drop each elements individually (in other words, only in the range that is not moved out). So we need ManuallyDrop on each element, and trasmuting should be the easiest so far.

**Comment by SimonSapin** [comment]
You can call `drop_in_place(…)` selectively on individual items or a sub-slice inside of `ManuallyDrop<[T; N]>`.

**Comment by ishitatsuyuki** [comment]
Good. I will update the text later.

File: text/0000-array_into_iter_move.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This is not 100% backwards compatible as it changes the signature of into_iter (only if directly
+called like below). A crater run is required.
+
+```rust
+[1, 2].into_iter();
+// This was originally yielding references, but now values.
+```
```
Comment (bluss) [inline text/0000-array_into_iter_move.md]: This is pretty much ensured to break actual code in many places, both in crater (that we can test) and lots of code we can't test. Maybe we can begin warning for this long before the new iterator is implemented?

Comment (ishitatsuyuki) [reply text/0000-array_into_iter_move.md]: Personally speaking, I don't think people would use `into_iter` despite the fact it yields references, and they would just `iter` instead. Thus I think fixing after a crater should be sufficient.

Comment (bluss) [reply text/0000-array_into_iter_move.md]: We don't reach all rust code in the world with crater. If the break is wide spread, we have to be more careful.

Comment (main--) [reply text/0000-array_into_iter_move.md]: And what about `for &x in [1,2,3]`, isn't that just `into_iter`?

Comment (bluss) [reply text/0000-array_into_iter_move.md]: That does not compile today I think so it's not a backwards company concern.

Comment (main--) [reply text/0000-array_into_iter_move.md]: Oh I see. `[1].into_iter()` currently resolves to the impl for `&[_; _]` which the new `[_; _]` impl would override, right?

**Comment by cristicbz** [comment]
@SimonSapin 

> I very much want this, but what is return type of [T; N]::into_iter? I think it should be IntoIter<T, N>, with const generics. So this needs to wait for that to be implemented.

This could be done without stabilising `Unsize`, by doing:
```rust
impl<T> Iterator for IntoIter<[T; 1]> { ... }
impl<T> Iterator for IntoIter<[T; 2]> { ... }
impl<T> Iterator for IntoIter<[T; 3]> { ... }
// ...
impl<T> Iterator for IntoIter<[T; 32]> { ... }
``` 

with all the impls calling to some shared unsafe method to avoid duplication. Has a nice migration story for when const generics land,  just replacing with

```rust
impl<T, const N> Iterator for IntoIter<[T; N]> {}
```

File: text/0000-array_into_iter_move.md (hunk 1 above)
Comment (bluss) [reply text/0000-array_into_iter_move.md]: Yes!

**Comment by ishitatsuyuki** [comment]
Please take a look, I think I have addressed the concerns.

**Comment by SimonSapin** [comment]
This looks great to me, thanks for writing it up!

**Comment by sinkuu** [comment]
[`Drop` impls cannot be specialized](https://github.com/rust-lang/rust/pull/23638), so we may have to use `Unsafe` or `FixedSizeArray` if we implement this RFC before const generics.

```rust
struct IntoIter<T> { ... }
impl<T> Drop for IntoIter<[T; 0]> { ... }
impl<T> Drop for IntoIter<[T; 1]> { ... } // error: Implementations of Drop cannot be specialized
...

// I expect this will work
struct IntoIter<T, const N: usize> { ... }
impl<T, const N: usize> Drop for IntoIter<T, N>

// This works today
struct IntoIter<T, A> where A: FixedSizeArray<T> { ... }
impl<T, A> Drop for IntoIter<T, A> where A: FixedSizeArray<T> { ... }
```

**Comment by eddyb** [comment]
@sinkuu What do you mean by `Unsafe`?

**Comment by ishitatsuyuki** [comment]
Presumably Unsize.

**Comment by ishitatsuyuki** [comment]
Should we make multiple IntoIter structs until const generic lands? Like this: `IntoIter32<T>`.

I think this should resolve the specialization problem, alongside reducing the template complexity, so we can store pointers (or unsafe slices) inside the struct.

**Comment by SimonSapin** [comment]
> `IntoIter32<T>`

I want this feature, but I don’t think I want to stabilize such an API.

It’d be harder to generalize once const generics land, too.

> so we can store pointers (or unsafe slices) inside the struct

I don’t understand what you mean here. The point of `into_iter` is that the items move when the iterator moves. Pointers should not be involved.

**Comment by ishitatsuyuki** [comment]
> but I don’t think I want to stabilize such an API.

I'm afraid that even `IntoIter<T>` has to change into `IntoIter<T, const N>` so signature change is unavoidable.

> I don’t understand what you mean here.

We need to save the range we have consumed with the iterator. It can be either usize-indexed or just start-end pointer.

**Comment by SimonSapin** [comment]
Changing the signature of stable APIs is not acceptable, so I’m afraid that this implies we should not stabilize anything here without const generics.

Self-referential pointers would become invalid when the iterator is moved. It has to be indices. (Two of them, or a `Range`, to support `DoubleEndedIterator`.)

**Comment by ishitatsuyuki** [comment]
I've come up with a partial stabilization plan, please let me know if this is possible.

As the signature of the iterator struct is subject to change, we will be keeping that struct(s) unstable until const generics land. I will call this struct `UnstableIntoIter` below.

`IntoIterator` is implemented like below, using the length of 32 as an example:

```rust
impl<T> IntoIterator for [T; 32] {
    type Item = T;
    type IntoIter = UnstableIntoIter32<T>;

    fn into_iter(self) -> T { ... }
}
```

I have no idea if this implementation will be accepted as stable in the compiler, so we can use it without the ability to annotate the type.

**Comment by kennytm** [comment]
Instead of a specific `IntoIter` or `UnstableIntoIter` method, could it just return an existential type `impl Iterator<Item=T>`?

**Comment by scottmcm** [comment]
Would it be compatible to replace
```rust
struct IntoIter0<T>{ .. };
..
struct IntoIter32<T>{ .. };
```
with
```rust
type IntoIter0<T> = IntoIter<T, const 0>;
..
type IntoIter32<T> = IntoIter<T, const 32>;
```
once const generics land?

Then the eventual type would be [the nice one](https://github.com/rust-lang/rfcs/pull/2185#issuecomment-338375424), and the residual cruft is just a bunch of deprecated type aliases, which should be easy to maintain and ignore in the docs.

**Comment by bluss** [comment]
We can do one thing before we even start the implementation: Work on deprecating & future warning the existing `array_value.into_iter()` calls.

**Comment by ishitatsuyuki** [comment]
I experimented with an implementation of deprecation warning, but it seems that I can't avoid one particular false positive: https://play.rust-lang.org/?gist=8b69b43bc947bf112c0f20d0e58b3add&version=stable

Basically, the differences between the below cannot be detected without changing the signature:

```
a.into_iter();
(&a).into_iter();
```

**Comment by sfackler** [comment]
We discussed this at the @rust-lang/libs meeting today, and feel like this won't really be actionable until integer generics are implemented.

@rfcbot fcp postpone

**Comment by bluss** [comment]
@sfackler What do you think about starting to warn / soft deprecate `[1, 2, 3].into_iter()` in favour of `[1, 2, 3].iter()` ?

**Comment by ishitatsuyuki** [comment]
I have mentioned that if we don't stabilize the iterator types, we can implement it with macro generated iterators and partially stabilize the into_iter() function only.

**Comment by sfackler** [comment]
@bluss 

Not sure - we haven't really done anything like that before. @aturon thoughts?

@ishitatsuyuki 

I am personally not a fan of the current "up to size 32" approach, and don't really want to expand that further.

**Comment by ishitatsuyuki** [comment]
@sfackler - what we only can do currently is that "up to size 32" approach; it should not be a blocker to improve array usability.

I originally thought that the main concern is backwards compatibility, which needs a lint to be developed. As I emphasized above, const generics is not a blocker for this, just a wishlist item. As const generics is an approved RFC, we can expect the stabilization at the same time.

**Comment by aturon** [comment]
@bluss I'd be in favor.

**Comment by Centril** [comment]
Closing since FCP with the proposal to postpone is now complete.

**Pull request closed without merging**